compiler/AST/iterator.cpp

```diff
@@ -765,21 +937,45 @@ static void
 buildIncr(IteratorInfo* ii, BlockStmt* singleLoop) {
   BlockStmt* incrBody = new BlockStmt();
 
-  if (singleLoop != NULL && singleLoop->isCForLoop() == true) {
-    CForLoop*  cforLoop  = toCForLoop(singleLoop);
-    BlockStmt* incrBlock = NULL;
-
-    Symbol*    ic        = ii->advance->getFormal(1);
-    SymbolMap  map;
+  // See Notes #1 and #2.
+  SymbolMap map;
+  Symbol* advance_this = ii->advance->getFormal(1);
+  Symbol* my_this = ii->incr->getFormal(1);
+  INT_ASSERT(my_this == ii->incr->_this);
+  map.put(advance_this, my_this);
 
-    // In copied expressions, replace _ic with incr->_this .
-    map.put(ic, ii->incr->_this);
+  if (singleLoop != NULL) {
+    if (singleLoop->isCForLoop() == true) {
+      CForLoop*  cforLoop  = toCForLoop(singleLoop);
+      BlockStmt* incrBlock = NULL;
 
-    incrBlock = cforLoop->incrBlockGet();
+      incrBlock = cforLoop->incrBlockGet();
 
-    for_alist(expr, incrBlock->body) {
-      incrBody->insertAtTail(expr->copy(&map));
+      for_alist(expr, incrBlock->body) {
+        incrBody->insertAtTail(expr->copy(&map));
+      }
     }
+    else if (singleLoop->isForLoop())
```

Comment:
Just to clarify. buildIncr was explicitly added when CForLoops were added. CForLoops are basically 

`CForLoop(init, test, incr) { }`. No other loop construct has an incr clause, it all just occurs in or before the loop body and gets included in the buildZip* segments